File: nodes/core/market/utils/extract_symbols_node.py
import logging
from typing import Any

from core.types_registry import AssetSymbol, OHLCVBar, get_type
from nodes.base.base_node import Base

logger = logging.getLogger(__name__)


class ExtractSymbols(Base):
    """
    Extracts a list of asset symbols from an OHLCV bundle.
    Takes an OHLCV bundle (Dict[AssetSymbol, List[OHLCVBar]]) and outputs
    just the list of AssetSymbol keys.
    """

    inputs = {"ohlcv_bundle": get_type("OHLCVBundle")}
    outputs = {"symbols": get_type("AssetSymbolList")}
    default_params = {}
    params_meta = []

    async def _execute_impl(self, inputs: dict[str, Any]) -> dict[str, list[AssetSymbol]]:
        ohlcv_bundle: dict[AssetSymbol, list[OHLCVBar]] = inputs.get("ohlcv_bundle", {})

        # Extract the asset symbols (keys) from the OHLCV bundle
        symbols = list(ohlcv_bundle.keys())

        return {"symbols": symbols}

File: backend/src/nodes/core/market/utils/extract-symbols-node.ts
// src/nodes/core/market/utils/extract-symbols-node.ts
// Translated from: nodes/core/market/utils/extract_symbols_node.py

import { Base } from '../../../base/base-node';
import { getType } from '../../../../core/types';
import type {
  NodeInputs,
  NodeOutputs,
  ParamMeta,
  DefaultParams,
  AssetSymbolData,
  OHLCVBar,
  OHLCVBundle,
} from '../../../../core/types';
import { AssetSymbol } from '../../../../core/types';

/**
 * Extracts a list of asset symbols from an OHLCV bundle.
 * Takes an OHLCV bundle (Map<AssetSymbol, OHLCVBar[]>) and outputs
 * just the list of AssetSymbol keys.
 */
export class ExtractSymbols extends Base {
  static override inputs: Record<string, unknown> = {
    ohlcv_bundle: getType('OHLCVBundle'),
  };
  static override outputs: Record<string, unknown> = {
    symbols: getType('AssetSymbolList'),
  };
  static override defaultParams: DefaultParams = {};
  static override paramsMeta: ParamMeta[] = [];

  protected override async executeImpl(inputs: NodeInputs): Promise<NodeOutputs> {
    const ohlcvBundle = (inputs.ohlcv_bundle as OHLCVBundle) ?? new Map();

    // Extract the asset symbols (keys) from the OHLCV bundle
    const symbols = Array.from(ohlcvBundle.keys());

    return { symbols };
  }
}
